src/library_async.js

```diff
@@ -823,6 +823,148 @@ mergeInto(LibraryManager.library, {
     });
   },
 
+  /*
+   * Layout of an ASYNCIFY Fiber structure
+   *
+   *  0 Asyncify context (Asyncify.currData)
+   *  4 STACKTOP
+   *  8 STACK_MAX
+   * 12 STACK_BASE
+   */
+  $Fibers: {
+    swapCounter: 0,     // HACK; see emscripten_fiber_swap
+    continuations: {},  // fiber pointer -> JS continuation function mapping
+  },
+
+  emscripten_fiber_create__sig: 'iiiii',
+  emscripten_fiber_create__deps: ['$Fibers', 'malloc'],
+  emscripten_fiber_create: function(funcptr, userdata, stack, stack_size) {
+    var fib = _malloc(16);
+
+    // init Asyncify context to null
+    {{{ makeSetValue('fib', 0, 0, 'i32') }}};
+
+    // stack grows down on the WASM backend, so base is at the top
+    var fstack_base = stack + stack_size;
+    var fstack_max = stack;
+
+    // init stack pointers
+    {{{ makeSetValue('fib',  4, 'fstack_base', 'i32') }}};
+    {{{ makeSetValue('fib',  8, 'fstack_max',  'i32') }}};
+    {{{ makeSetValue('fib', 12, 'fstack_base', 'i32') }}};
+
+    // init continuation (call user-supplied function on first entry)
+    Fibers.continuations[fib] = () => {
+#if STACK_OVERFLOW_CHECK
+      writeStackCookie();
+#endif
+      {{{ makeDynCall('vi') }}}(funcptr, userdata);
+    };
+
+    return fib;
+  },
+
+  emscripten_fiber_create_from_current_context__sig: 'i',
+  emscripten_fiber_create_from_current_context__deps: ['$Fibers', 'malloc'],
+  emscripten_fiber_create_from_current_context: function() {
+    var fib = _malloc(16);
+
+    {{{ makeSetValue('fib',  8, 'STACK_MAX',  'i32') }}};
+    {{{ makeSetValue('fib', 12, 'STACK_BASE', 'i32') }}};
+
+    // NOTE: We don't need to initialize the other fields, because swapping will
+    // overwrite them anyway. We can not swap into this fiber without swapping
+    // out of it first anyway, because we'd need a continuation function.
+    //
+    // It may be possible to support this by doing an Asyncify.handleSleep here
+    // just to capture the wakeUp callback, but I'm not sure if it's a useful
+    // thing to support.
+
+    return fib;
+  },
+
+  emscripten_fiber_recycle__sig: 'viii',
+  emscripten_fiber_recycle__deps: ["$Fibers"],
+  emscripten_fiber_recycle: function(fib, funcptr, userdata) {
+    {{{ makeSetValue('fib', 0, 0, 'i32') }}};
+
+    var fstack_base = {{{ makeGetValue('fib', 12, 'i32') }}};
+    {{{ makeSetValue('fib', 4, 'fstack_base', 'i32') }}};
+
+    Fibers.continuations[fib] = () => {
```

Comment:
In general we don't use ES6 features like `=>` in emscripten JS, as we want the code to be runnable in older browsers.

(We may at some point use babel to support them, but that hasn't been done yet.)

Reply:
I don't have a problem with changing this, I'm just curious: does anything that supports wasm not have ES6 features by now? I know that there's wasm2js for old browsers; does it work with asyncify?

Reply:
We support non-wasm output too (using wasm2js, which emits JS). That gives people the option to support older browsers and VMs. Otherwise, yeah, when using actual wasm most (all?) ES6 would be ok.

Reply:
Oh, and yes, asyncify is supported with wasm2js.